Fix: optimize gitops doc (#308)
oam-dev/kubevela.io: docs/case-studies/gitops.md

In this section, we will introduce steps of using KubeVela directly in GitOps approach.

> There are two modes of delivery, which we will introduce separately:

Reviewer: ```suggestion
This article will separate into two perspectives:
```

> 1. For platform administrators/SREs, users can update the KubeVela configuration file in the repository directly, and the application in the cluster will be updated by KubeVela automatically.

Reviewer: ```suggestion
1. For platform administrators/SREs, they can update the  config in Git repo. It will trigger automated re-deployment.
```

> 2. For developers, users can update the code in the application code repository, and KubeVela will update the application in the cluster with the latest image automatically.

Reviewer: ```suggestion
2. For developers, they can update the app source code and then push it to Git. It will trigger building latest image and re-deployment.
```

> Note: you can also use it with existing tools such as ArgoCD with similar steps, detailed guides will be added in following releases.

## For platform administrators/SREs

![alt](../resources/ops-flow.jpg)

> For platform administrators/SREs, they only need to prepare a KubeVela Git configuration repository and apply the KubeVela files. Then they can update the config in the Git configuration repository directly, KubeVela will watch the configuration repository and update the cluster, making every configuration change traceable.

Reviewer: ```suggestion
Platform administrators/SREs prepares the Git repo for operational config. Every config config change will be traceable by that. KubeVela will watch the repo and apply changes to the clusters.
```

## Setup Config Repository

> The configuration files are from the [Example Repo](https://github.com/oam-dev/samples/tree/master/9.GitOps_Demo).

The structure of the config repository looks below:

> * The `clusters/` contains the KubeVela config in cluster. Users need to apply the files in this directory to cluster manually, KubeVela will then watch the repository and update the cluster.

Reviewer: ```suggestion
* The `clusters/` contains the GitOps config. It will command KubeVela to watch the specified repo and apply latest changes.
```

> * The `apps/` contains the configuration of the applicatiThe files will be watched by `app.yaml` in `clusters/`.on. 

Reviewer: ```suggestion
* The `apps/` contains the Application yaml for deploying the user-facing app.
```

> * The `infrastructure/` contains some infrastructure tools, such as MySQL database. The files will be watched by `infra.yaml` in `clusters/`.

Reviewer: ```suggestion
* The `infrastructure/` contains infrastructure tools, i.e. MySQL database.
```

```shell
├── apps
│   └── my-app.yaml
├── clusters
│   ├── apps.yaml
│   └── infra.yaml
└── infrastructure
    └── mysql.yaml
```

> KubeVela recommends using the directory structure above to manage your GitOps repository. `clusters/` holds the associated KubeVela configuration that need to be applied to cluster manually, `apps/` holds your application and `infrastructure/` holds your base configuration. Application-related configurations can be bound with traits. By separating applications from basic configurations, you can manage your deployment environment more reasonably and isolate application changes.

#### Directory `clusters/`

> `apps.yaml` is almost the same as `infra.yaml` in `clusters/`, the only difference is the watched directory.

Reviewer: ```suggestion
`apps.yaml` and `infra.yaml` in `clusters/` are similar. Their difference is to watch different directories.
```

> 

Reviewer: ```suggestion
```

> Apply the files in `clusters/` manually, they can pull and sync the files in `infrastructure/` and `apps/` automatically.

Reviewer: ```suggestion
Apply the files in `clusters/` manually. They will sync the files in `infrastructure/` and `apps/` dir of the Git repo.
```

Reviewer: 这句要不放到 yaml 例子后面？

Author: 之前的 apply xxx 都是写在 yaml 前的吧

```yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:

>   name: infra

Reviewer: 这个 yaml 前面得再加一段话：

Below is how the `infra.yaml` looks like:

> Note: you can also use it with existing tools such as ArgoCD with similar steps, detailed guides will be added in following releases.

> ## Setup Code Repository

Reviewer: ```suggestion
## Setup App Code Repository
```

> First, setup a Git code repository with some source code and a Dockerfile.

Reviewer: ```suggestion
First, setup a Git repository with source code and Dockerfile.
```

> The code is very simple, starting a service and connect to a MySQL database. In the default path, it will display the version in the code; in the `/db` path, it will list the data in database.

Reviewer: ```suggestion
The deployment looks like this: it starts a service and connects it to a MySQL database. In the default path, it will display the version in the code; in the `/db` path, it will list the data in database.
```
```suggestion
The app serves HTTP service and connects to a MySQL database. In the '/' path, it will display the version in the code; in the `/db` path, it will list the data in database.
```

```go
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		_, _ = fmt.Fprintf(w, "Version: %s\n", VERSION)
	})
	http.HandleFunc("/db", func(w http.ResponseWriter, r *http.Request) {
		rows, err := db.Query("select * from userinfo;")
		if err != nil {
			_, _ = fmt.Fprintf(w, "Error: %v\n", err)
		}
		for rows.Next() {
			var username string
			var desc string
			err = rows.Scan(&username, &desc)
			if err != nil {
				_, _ = fmt.Fprintf(w, "Scan Error: %v\n", err)
			}
			_, _ = fmt.Fprintf(w, "User: %s \nDescription: %s\n\n", username, desc)
		}
	})

	if err := http.ListenAndServe(":8088", nil); err != nil {
		panic(err.Error())
	}
```

> We want users to build the image and push it to the image registry after changing the code, so we need to integrate with a CI tool like GitHub Actions or Jenkins to do it. In this example, we use GitHub Actions to build the image. For the code and configuration file, please refer to [Example Repo](https://github.com/oam-dev/samples/tree/master/9.GitOps_Demo).

Reviewer: ```suggestion
 In this tutorial, we will setup a CI pipeline using GitHub Actions to build the image and push it to a registry. The code and configuration files are from the [Example Repo](https://github.com/oam-dev/samples/tree/master/9.GitOps_Demo).
```

> ## Create the Git secret

Reviewer: ```suggestion
## Create Git Secret for KubeVela committing to Config Repo
```

> After the new image is pushed to the image registry, KubeVela will recognize the new image and update the `Application` file in the Git repository and cluster. Therefore, we need a secret with Git information for KubeVela to commit to the Git repository. Fill the following yaml files with your password and apply it to the cluster:

Reviewer: ```suggestion
After the new image is pushed to the image registry, KubeVela will be notified and update the `Application` file in the Git repository and cluster. Therefore, we need a secret with Git information for KubeVela to commit to the Git repository. Fill the following yaml files with your password and apply it to the cluster:
```

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: my-secret
type: kubernetes.io/basic-auth
stringData:
  username: <your username>
  password: <your password>
```

## Setup Config Repository

> We also need a config repository for clusters, applications, and some infrastructure configuration.

Reviewer: ```suggestion
We need a config repository for applications and infrastructure configuration.
```

> The structure of the config repository is as bellow:

Reviewer: ```suggestion
The structure of the config repository looks below:
```

              /: 8088
```

#### Directory `infrastructure/`

> The `infrastructure/` contains the config of some infrastructures like database.In this case, we use [mysql controller](https://github.com/bitpoke/mysql-operator) to deploy a MySQL cluster.

Reviewer: ```suggestion
The `infrastructure/` contains the config of some infrastructures like database. In the following, we will use [MySQL operator](https://github.com/bitpoke/mysql-operator) to deploy a MySQL cluster.
```

```

`curl` the Ingress of the app, we can see that the current version is `0.1.5` and the application is connected to the database successfully:

```shell
$ kubectl get ingress
NAME        CLASS    HOSTS                 ADDRESS         PORTS   AGE
my-server   <none>   testsvc.example.com   <ingress-ip>    80      162m

$ curl -H "Host:testsvc.example.com" http://<ingress-ip>
Version: 0.1.5

$ curl -H "Host:testsvc.example.com" http://<ingress-ip>/db
User: KubeVela
Description: It's a test user
```

## Modify the config for GitOps trigger

After the first deployment, we can modify the files in config repo to update the applications in the cluster.

Modify the domain of the application's Ingress:

```yaml
...
      traits:
        - type: ingress
          properties:
            domain: kubevela.example.com
            http:
              /: 8089
```

Check the Ingress in cluster after a while:

```shell
NAME        CLASS    HOSTS                 ADDRESS         PORTS   AGE
my-server   <none>   kubevela.example.com  <ingress-ip>    80      162m
```

The host of the Ingress has been updated successfully!

## For developers

![alt](../resources/dev-flow.jpg)

> For developers, they need to prepare an application code repository in addition to the KubeVela configuration repository. After the user updates the code in the application repository, a CI needs to be configured to automatically build the image and push it to the image registry. KubeVela listens for the latest images in the registry, automatically updates the image configuration in the repository, and finally updates the application configuration in the cluster.

Reviewer: ```suggestion
Developers writes the application source code and push it to a Git repo (aka app repo). Once app repo updates, the CI will build the image and push it to the image registry. KubeVela watches the image registry, and updates the image in config repo. Finally, it will apply the config to the cluster.
```

    properties:
      repoType: git
      # replace it with your repo url
      url: <your repo url>
      # replace it with your git secret
      secretRef: git-secret
      # the pull interval time, set to 10m since the infrastructure is steady
      pullInterval: 10m
      git:
        # the branch name
        branch: main
      # the path to sync
      path: ./infrastructure
```

Apply the above file to the clusters, we can see that the MySQL in `infrastructure/` is automatically deployed:

```shell
$ vela ls

APP   	COMPONENT       	TYPE      	TRAITS 	PHASE  	HEALTHY	STATUS	CREATED-TIME
infra 	database-config 	kustomize 	       	running	healthy	      	2021-09-26 20:48:09 +0800 CST
mysql 	mysql-controller	helm      	       	running	healthy	      	2021-09-26 20:48:11 +0800 CST
└─  	mysql-cluster   	raw       	       	running	healthy	      	2021-09-26 20:48:11 +0800 CST
```

Check the config of application, this file will pull and sync the files in `apps/` and watch the image update in registry:

```yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: apps
spec:
  components:
  - name: apps
    type: kustomize
    properties:
      repoType: git

>       url: <your github repo address>

Reviewer: What's this github repo address? The app or the config repo?

        commitMessage: "Image: {{range .Updated.Images}}{{println .}}{{end}}"
```

Apply the above file to the clusters, we can see that the application in `apps/` is automatically deployed:

```shell
APP   	COMPONENT       	TYPE      	TRAITS 	PHASE  	HEALTHY	STATUS	CREATED-TIME
apps  	apps            	kustomize 	       	running	healthy	      	2021-09-27 16:55:53 +0800 CST
infra 	database-config 	kustomize 	       	running	healthy	      	2021-09-26 20:48:09 +0800 CST
my-app	my-server       	webservice	ingress	running	healthy	      	2021-09-27 16:55:55 +0800 CST
mysql 	mysql-controller	helm      	       	running	healthy	      	2021-09-26 20:48:11 +0800 CST
└─  	mysql-cluster   	raw       	       	running	healthy	      	2021-09-26 20:48:11 +0800 CST
```

> Curl the Ingress of the app, we can see that the current version is `0.1.5` and the application is connected to the database successfully:

Reviewer: ```suggestion
`curl` the Ingress of the app, we can see that the current version is `0.1.5` and the application is connected to the database successfully:
```

```shell
$ kubectl get ingress
NAME        CLASS    HOSTS                 ADDRESS         PORTS   AGE
my-server   <none>   testsvc.example.com   <ingress-ip>    80      162m

$ curl -H "Host:testsvc.example.com" http://<ingress-ip>
Version: 0.1.5

$ curl -H "Host:testsvc.example.com" http://<ingress-ip>/db
User: KubeVela
Description: It's a test user
```

> ## Modify the configuration

Reviewer: ```suggestion
## Modify the config for GitOps trigger
```

After the first deployment, we can modify the files in config repo to update the applications in the cluster.

Modify the domain of the application's Ingress:

```yaml
...
      traits:
        - type: ingress
          properties:
            domain: kubevela.example.com
            http:
              /: 8089
```

> Check the Ingress in cluster after a while:

Reviewer: Who is  listening to the config repo and apply to the cluster?
IIRC the previous sections have only mentioned how to make KubeVela committing to the Code Repo

```shell
NAME        CLASS    HOSTS                 ADDRESS         PORTS   AGE
my-server   <none>   kubevela.example.com  <ingress-ip>    80      162m
```

The host of the Ingress has been updated successfully!

## Modify the code

In addition to modify the configuration in the repository directly, we can also modify the code in the repository to the upgrade and apply the application automatically.

Change the `Version` to `0.1.6` and modify the data in database:

```go
const VERSION = "0.1.6"

...

func InsertInitData(db *sql.DB) {
	stmt, err := db.Prepare(insertInitData)
	if err != nil {
		panic(err)
	}
	defer stmt.Close()

	_, err = stmt.Exec("KubeVela2", "It's another test user")
	if err != nil {
		panic(err)
	}
}
```

Commit the change to the Git Repository, we can see that our CI pipelines has built the image and push it to the image registry.

> KubeVela will then listening to the image registry and update the `Application` in Git Repository with the latest image tag. Notice that KubeVela will only be able to modify the image field if the annotation `# {"$imagepolicy": "default:apps"}` is added after the field. `default:apps` is `namespace:application-name` of the application config file.

Reviewer: ```suggestion
KubeVela will listen to the image registry and update the `Application` in Git Repository with the latest image tag. Notice that KubeVela will only be able to modify the image field if the annotation `# {"$imagepolicy": "default:apps"}` is added after the field. `default:apps` is `namespace:application-name` of the application config file.
```

> * When the `Application` file in the Git repository is updated, KubeVela will update the `Application` in the cluster based on the latest configuration.
> * When a new tag is added to the image registry, KubeVela will filter out the latest tag based on your policy and update it to Git repository. When the files in the repository are updated, KubeVela repeats the first step and updates the files in the cluster, thus achieving automatic deployment.

We can `curl` to `Ingress` to see the current version and data:

```shell
$ kubectl get ingress
NAME        CLASS    HOSTS                 ADDRESS         PORTS   AGE
my-server   <none>   kubevela.example.com  <ingress-ip>    80      162m

$ curl -H "Host:kubevela.example.com" http://<ingress-ip>
Version: 0.1.6

$ curl -H "Host:kubevela.example.com" http://<ingress-ip>/db
User: KubeVela
Description: It's a test user

User: KubeVela2
Description: It's another test user
```

The `Version` has been updated successfully! Now we're done with everything from changing the code to automatically applying to the cluster.

## Summary

> On the development side, KubeVela automatically updates the image in the configuration repository after the user modifies the code in the repository, make it easier for developers to deploy and update their apps.

Reviewer: ```suggestion
For end user developers, KubeVela automatically updates the image in the configuration repository after the user modifies the code in the repository, make it easier for developers to deploy and update their apps.
```

> On the operation side, if you need to update the configuration of the infrastructure (such as database) or other fields of the application, the only thing you need to do is to modify the files in the configuration repository. KubeVela will automatically synchronize the configuration to the cluster, thus simplifying the deployment process.

Reviewer: ```suggestion
For platform admin/SRE, to update the configuration of the infrastructure (such as database) or other fields of the application, the only thing you need to do is to modify the files in the configuration repository. KubeVela will automatically synchronize the configuration to the cluster, thus simplifying the deployment process.
```

> * When the `Application` file in the Git repository is updated, KubeVela will update the `Application` in the cluster based on the latest configuration.
> * When a new tag is added to the image registry, KubeVela will filter out the latest tag based on your policy and update it to Git repository. When the files in the repository are updated, KubeVela repeats the first step and updates the files in the cluster, thus achieving automatic deployment.

We can `curl` to `Ingress` to see the current version and data:

```shell
$ kubectl get ingress
NAME        CLASS    HOSTS                 ADDRESS         PORTS   AGE
my-server   <none>   kubevela.example.com  <ingress-ip>    80      162m

$ curl -H "Host:kubevela.example.com" http://<ingress-ip>
Version: 0.1.6

$ curl -H "Host:kubevela.example.com" http://<ingress-ip>/db
User: KubeVela
Description: It's a test user

User: KubeVela2
Description: It's another test user
```

The `Version` has been updated successfully! Now we're done with everything from changing the code to automatically applying to the cluster.

## Summary

> For platform admin/SRE, to update the configuration of the infrastructure (such as database) or other fields of the application, the only thing you need to do is to modify the files in the configuration repository. KubeVela will automatically synchronize the configuration to the cluster, thus simplifying the deployment process.

Reviewer: ```suggestion
For platform admins/SREs, they update the config repo to operate the application and infrastructure. KubeVela will synchronize the config to the cluster, simplifying the deployment process.
```

> For end user developers, KubeVela automatically updates the image in the configuration repository after the user modifies the code in the repository, make it easier for developers to deploy and update their apps.

Reviewer: ```suggestion
For end users/developers, they write the source code, push it to Git, and then re-deployment will happen. It will make CI to build the image. KubeVela will then update the image field and apply the deployment config.
```
